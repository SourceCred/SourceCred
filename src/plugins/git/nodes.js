// @flow

import {NodeAddress, type NodeAddressT} from "../../core/graph";
import type {Hash} from "./types";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GIT_PREFIX = NodeAddress.fromParts(["sourcecred", "git"]);
export function _gitAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GIT_PREFIX, ...parts);
}

export const BLOB_TYPE: "BLOB" = "BLOB";
export const COMMIT_TYPE: "COMMIT" = "COMMIT";
export const TREE_TYPE: "TREE" = "TREE";
export const TREE_ENTRY_TYPE: "TREE_ENTRY" = "TREE_ENTRY";

export const _Prefix = Object.freeze({
  base: GIT_PREFIX,
  blob: _gitAddress(BLOB_TYPE),
  commit: _gitAddress(COMMIT_TYPE),
  tree: _gitAddress(TREE_TYPE),
  treeEntry: _gitAddress(TREE_ENTRY_TYPE),
});

export type BlobAddress = {|
  +type: typeof BLOB_TYPE,
  +hash: Hash,
|};
export type CommitAddress = {|
  +type: typeof COMMIT_TYPE,
  +hash: Hash,
|};
export type TreeAddress = {|
  +type: typeof TREE_TYPE,
  +hash: Hash,
|};
export type TreeEntryAddress = {|
  +type: typeof TREE_ENTRY_TYPE,
  +treeHash: Hash,
  +name: string,
|};

// A tree entry has contents with one of the following types of
// addresses.
export type TreeEntryContentsAddress =
  | BlobAddress
  | CommitAddress
  | TreeAddress;

export type StructuredAddress =
  | BlobAddress
  | CommitAddress
  | TreeAddress
  | TreeEntryAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = NodeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "BLOB": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: BLOB_TYPE, hash};
    }
    case "COMMIT": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: COMMIT_TYPE, hash};
    }
    case "TREE": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: TREE_TYPE, hash};
    }
    case "TREE_ENTRY": {
      if (rest.length !== 2) throw fail();
      const [treeHash, name] = rest;
      return {type: TREE_ENTRY_TYPE, treeHash, name};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case BLOB_TYPE:
      return NodeAddress.append(_Prefix.blob, x.hash);
    case COMMIT_TYPE:
      return NodeAddress.append(_Prefix.commit, x.hash);
    case TREE_TYPE:
      return NodeAddress.append(_Prefix.tree, x.hash);
    case TREE_ENTRY_TYPE:
      return NodeAddress.append(_Prefix.treeEntry, x.treeHash, x.name);
    default:
      // eslint-disable-next-line no-unused-expressions
      (x.type: empty);
      throw new Error(`Unexpected type ${x.type}`);
  }
}
