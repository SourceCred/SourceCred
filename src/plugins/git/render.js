// @flow

import type {Address} from "core/address";
import type {Graph} from "core/graph";
import type {NodeType, SubmoduleCommitPayload} from "./types";

/**
 * Describe a node provided by the Git plugin.
 */
export function nodeDescription(graph: Graph<any, any>, address: Address) {
  const type: NodeType = (address.type: any);
  switch (type) {
    case "COMMIT":
      return `commit ${address.id}`;
    case "TREE":
      return `tree ${address.id}`;
    case "BLOB":
      return `blob ${address.id}`;
    case "SUBMODULE_COMMIT": {
      const payload: SubmoduleCommitPayload = graph.node(address).payload;
      return `submodule commit ${payload.hash} in ${payload.url}`;
    }
    case "TREE_ENTRY":
      return `entry ${address.id}`;
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw new Error(`unknown type: ${type}`);
  }
}
