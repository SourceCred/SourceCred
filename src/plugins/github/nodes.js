// @flow

import {NodeAddress, type NodeAddressT} from "../../core/graph";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GITHUB_PREFIX = NodeAddress.fromParts(["sourcecred", "github"]);
export function _githubAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GITHUB_PREFIX, ...parts);
}

export const REPO_TYPE: "REPO" = "REPO";
export const ISSUE_TYPE: "ISSUE" = "ISSUE";
export const PULL_TYPE: "PULL" = "PULL";
export const REVIEW_TYPE: "REVIEW" = "REVIEW";
export const COMMENT_TYPE: "COMMENT" = "COMMENT";
export const USERLIKE_TYPE: "USERLIKE" = "USERLIKE";

export const _Prefix = Object.freeze({
  base: GITHUB_PREFIX,
  repo: _githubAddress(REPO_TYPE),
  issue: _githubAddress(ISSUE_TYPE),
  pull: _githubAddress(PULL_TYPE),
  review: _githubAddress(REVIEW_TYPE),
  comment: _githubAddress(COMMENT_TYPE),
  userlike: _githubAddress(USERLIKE_TYPE),
  reviewComment: _githubAddress(COMMENT_TYPE, REVIEW_TYPE),
  issueComment: _githubAddress(COMMENT_TYPE, ISSUE_TYPE),
  pullComment: _githubAddress(COMMENT_TYPE, PULL_TYPE),
});

export type RepoAddress = {|
  +type: typeof REPO_TYPE,
  +owner: string,
  +name: string,
|};
export type IssueAddress = {|
  +type: typeof ISSUE_TYPE,
  +repo: RepoAddress,
  +number: string,
|};
export type PullAddress = {|
  +type: typeof PULL_TYPE,
  +repo: RepoAddress,
  +number: string,
|};
export type ReviewAddress = {|
  +type: typeof REVIEW_TYPE,
  +pull: PullAddress,
  +id: string,
|};
export type CommentAddress = {|
  +type: typeof COMMENT_TYPE,
  +parent: CommentableAddress,
  +id: string,
|};
export type UserlikeAddress = {|
  +type: typeof USERLIKE_TYPE,
  +login: string,
|};

export type StructuredAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | UserlikeAddress;

// Each of these types has 0 or more "AUTHORS" edges, each of which
// leads to a UserlikeAddress.  Note: It is not true that every
// Authorable has at least one author, as when GitHub accounts are
// deleted, they leave behind posts without authors.
export type AuthorableAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress;

// Each of these types has text content, which means
// it may be the source of a reference to a ReferentAddress.
export type TextContentAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress;

// Each of these types may be referred to by something
// with text content.
export type ReferentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | UserlikeAddress;

// Each of these types is structurally the child of some
// entity with a ParentAddress.
export type ChildAddress =
  | IssueAddress // child of RepoAddress
  | PullAddress // child of RepoAddress
  | ReviewAddress // child of PullAddress
  | CommentAddress; // child of IssueAddress, PullAddress, or ReviewAddress

// Each of these types may have child nodes (see ChildAddress).
export type ParentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress;

// Each of these types may have Comments as children
export type CommentableAddress = IssueAddress | PullAddress | ReviewAddress;

// Verify that Commentable is a subtype of Parent
const _unused_static = (_: CommentableAddress): ParentAddress => _;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GITHUB_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_gh, kind, ...rest] = NodeAddress.toParts(x);
  switch (kind) {
    case REPO_TYPE: {
      if (rest.length !== 2) {
        throw fail();
      }
      const [owner, name] = rest;
      return {type: REPO_TYPE, owner, name};
    }
    case ISSUE_TYPE: {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: REPO_TYPE, owner, name};
      return {type: ISSUE_TYPE, repo, number};
    }
    case PULL_TYPE: {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: REPO_TYPE, owner, name};
      return {type: PULL_TYPE, repo, number};
    }
    case REVIEW_TYPE: {
      if (rest.length !== 4) {
        throw fail();
      }
      const [owner, name, pullNumber, id] = rest;
      const repo = {type: REPO_TYPE, owner, name};
      const pull = {type: PULL_TYPE, repo, number: pullNumber};
      return {type: REVIEW_TYPE, pull, id};
    }
    case COMMENT_TYPE: {
      if (rest.length < 1) {
        throw fail();
      }
      const [subkind, ...subrest] = rest;
      switch (subkind) {
        case ISSUE_TYPE: {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, issueNumber, id] = subrest;
          const repo = {type: REPO_TYPE, owner, name};
          const issue = {type: ISSUE_TYPE, repo, number: issueNumber};
          return {type: COMMENT_TYPE, parent: issue, id};
        }
        case PULL_TYPE: {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, pullNumber, id] = subrest;
          const repo = {type: REPO_TYPE, owner, name};
          const pull = {type: PULL_TYPE, repo, number: pullNumber};
          return {type: COMMENT_TYPE, parent: pull, id};
        }
        case REVIEW_TYPE: {
          if (subrest.length !== 5) {
            throw fail();
          }
          const [owner, name, pullNumber, reviewFragment, id] = subrest;
          const repo = {type: REPO_TYPE, owner, name};
          const pull = {type: PULL_TYPE, repo, number: pullNumber};
          const review = {type: REVIEW_TYPE, pull, id: reviewFragment};
          return {type: COMMENT_TYPE, parent: review, id};
        }
        default:
          throw fail();
      }
    }
    case USERLIKE_TYPE: {
      if (rest.length !== 1) {
        throw fail();
      }
      const [login] = rest;
      return {type: USERLIKE_TYPE, login};
    }
    default:
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case REPO_TYPE:
      return NodeAddress.append(_Prefix.repo, x.owner, x.name);
    case ISSUE_TYPE:
      return NodeAddress.append(
        _Prefix.issue,
        x.repo.owner,
        x.repo.name,
        x.number
      );
    case PULL_TYPE:
      return NodeAddress.append(
        _Prefix.pull,
        x.repo.owner,
        x.repo.name,
        x.number
      );
    case REVIEW_TYPE:
      return NodeAddress.append(
        _Prefix.review,
        x.pull.repo.owner,
        x.pull.repo.name,
        x.pull.number,
        x.id
      );
    case COMMENT_TYPE:
      switch (x.parent.type) {
        case ISSUE_TYPE:
          return NodeAddress.append(
            _Prefix.issueComment,
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.id
          );
        case PULL_TYPE:
          return NodeAddress.append(
            _Prefix.pullComment,
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.id
          );
        case REVIEW_TYPE:
          return NodeAddress.append(
            _Prefix.reviewComment,
            x.parent.pull.repo.owner,
            x.parent.pull.repo.name,
            x.parent.pull.number,
            x.parent.id,
            x.id
          );
        default:
          throw new Error(`Bad comment parent type: ${(x.parent.type: empty)}`);
      }
    case USERLIKE_TYPE:
      return NodeAddress.append(_Prefix.userlike, x.login);
    default:
      throw new Error(`Unexpected type ${(x.type: empty)}`);
  }
}
