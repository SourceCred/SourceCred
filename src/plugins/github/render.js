// @flow

import React, {type Node as ReactNode} from "react";
import * as R from "./relationalView";

function EntityUrl(props) {
  return (
    <a href={props.entity.url()} target="_blank" rel="nofollow noopener">
      {props.children}
    </a>
  );
}

function repo(x: R.Repo) {
  return (
    <EntityUrl entity={x}>
      {x.owner()}/{x.name()}
    </EntityUrl>
  );
}

function hyperlinkedNumber(x: R.Issue | R.Pull) {
  return <EntityUrl entity={x}>#{x.number()}</EntityUrl>;
}

function issue(x: R.Issue) {
  return (
    <span>
      {hyperlinkedNumber(x)}: {x.title()}
    </span>
  );
}

function pull(x: R.Pull) {
  const additions = <span style={{color: "green"}}>+{x.additions()}</span>;
  const deletions = <span style={{color: "red"}}>âˆ’{x.deletions()}</span>;
  const diff = (
    <span>
      ({additions}/{deletions})
    </span>
  );
  return (
    <span>
      {hyperlinkedNumber(x)} {diff}: {x.title()}
    </span>
  );
}

function review(x: R.Review) {
  const leader = <EntityUrl entity={x}>review</EntityUrl>;
  return (
    <span>
      {leader} on {description(x.parent())}
    </span>
  );
}

function comment(x: R.Comment) {
  const leader = <EntityUrl entity={x}>comment</EntityUrl>;
  return (
    <span>
      {leader} on {description(x.parent())}
    </span>
  );
}

function userlike(x: R.Userlike) {
  return <EntityUrl entity={x}>@{x.login()}</EntityUrl>;
}

// The commit type is included for completeness's sake and to
// satisfy the typechecker, but won't ever be seen in the frontend
// because the commit has a Git plugin prefix and will therefore by
// handled by the git plugin adapter
function commit(x: R.Commit) {
  // TODO(@wchargin): Ensure this hash is unambiguous
  const shortHash = x.address().hash.slice(0, 7);
  return (
    <span>
      Commit <EntityUrl entity={x}>{shortHash}</EntityUrl>
    </span>
  );
}

export function description(e: R.Entity): ReactNode {
  const handlers = {
    repo,
    issue,
    pull,
    review,
    comment,
    commit,
    userlike,
  };
  return R.match(handlers, e);
}
