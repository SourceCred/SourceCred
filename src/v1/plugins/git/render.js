// @flow

import {NodeReference} from "../../core/porcelain";
import {SubmoduleCommitReference, GitReference} from "./porcelain";

/**
 * Describe a node provided by the Git plugin.
 */
export function nodeDescription(nodeReference: NodeReference<any>) {
  const gReference = new GitReference(nodeReference);
  const type = gReference.type();
  const address = gReference.address();
  switch (type) {
    case "COMMIT":
      return `commit ${address.id}`;
    case "TREE":
      return `tree ${address.id}`;
    case "BLOB":
      return `blob ${address.id}`;
    case "SUBMODULE_COMMIT": {
      const scRef = new SubmoduleCommitReference(gReference);
      const scPorcelain = scRef.get();
      if (scPorcelain != null) {
        return `submodule commit ${scPorcelain.hash()} in ${scPorcelain.url()}`;
      } else {
        return `submodule commit [unknown]`;
      }
    }
    case "TREE_ENTRY":
      return `entry ${address.id}`;
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw new Error(`unknown type: ${type}`);
  }
}
