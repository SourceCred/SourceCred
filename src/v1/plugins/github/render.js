// @flow

/*
 * Methods for rendering and displaying GitHub nodes.
 */
import stringify from "json-stable-stringify";
import type {NodeReference} from "../../core/porcelain";
import type {NodePayload} from "./types";
import {
  GithubReference,
  AuthorReference,
  AuthorPorcelain,
  CommentReference,
  IssueReference,
  IssuePorcelain,
  PullRequestReference,
  PullRequestPorcelain,
  PullRequestReviewReference,
  PullRequestReviewCommentReference,
  RepositoryPorcelain,
} from "./porcelain";

/* Give a short description for the GitHub node at given address.
 * Useful for e.g. displaying a title.
 */
export function nodeDescription(ref: NodeReference<NodePayload>) {
  const porcelain = ref.get();
  if (porcelain == null) {
    return `[unknown ${ref.type()}]`;
  }
  const type = new GithubReference(ref).type();
  switch (type) {
    case "REPOSITORY": {
      const repo = new RepositoryPorcelain(porcelain);
      return `${repo.owner()}/${repo.name()}`;
    }
    case "ISSUE": {
      const issue = new IssuePorcelain(porcelain);
      return `#${issue.number()}: ${issue.title()}`;
    }
    case "PULL_REQUEST": {
      const pr = new PullRequestPorcelain(porcelain);
      const diff = `+${pr.additions()}/\u2212${pr.deletions()}`;
      return `#${pr.number()} (${diff}): ${pr.title()}`;
    }
    case "COMMENT": {
      const comment = new CommentReference(ref);
      const issue = comment.parent();
      return `comment by @${authors(comment)} on #${num(issue)}`;
    }
    case "PULL_REQUEST_REVIEW": {
      const review = new PullRequestReviewReference(ref);
      const pr = review.parent();
      return `review by @${authors(review)} on #${num(pr)}`;
    }
    case "PULL_REQUEST_REVIEW_COMMENT": {
      const comment = new PullRequestReviewCommentReference(ref);
      const pr = comment.parent().parent();
      return `review comment by @${authors(comment)} on #${num(pr)}`;
    }
    case "AUTHOR": {
      return `@${new AuthorPorcelain(porcelain).login()}`;
    }
    default: {
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw new Error(
        `Tried to write description for invalid type ${stringify(
          ref.address()
        )}`
      );
    }
  }
}

function num(x: IssueReference | PullRequestReference) {
  const np = x.get();
  return np == null ? "[unknown]" : np.number();
}

function authors(authorable: {+authors: () => AuthorReference[]}) {
  // TODO: modify to accomodate multi-authorship
  const authorRefs = authorable.authors();
  const firstAuthor = authorRefs[0].get();
  return firstAuthor != null ? firstAuthor.login() : "[unknown]";
}
