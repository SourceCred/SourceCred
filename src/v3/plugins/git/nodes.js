// @flow

import {NodeAddress, type NodeAddressT} from "../../core/graph";
import type {Hash} from "./types";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GIT_PREFIX = NodeAddress.fromParts(["sourcecred", "git"]);
export function _gitAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GIT_PREFIX, ...parts);
}

export const BLOB_TYPE: "BLOB" = "BLOB";
export const COMMIT_TYPE: "COMMIT" = "COMMIT";
export const SUBMODULE_COMMIT_TYPE: "SUBMODULE_COMMIT" = "SUBMODULE_COMMIT";
export const TREE_TYPE: "TREE" = "TREE";
export const TREE_ENTRY_TYPE: "TREE_ENTRY" = "TREE_ENTRY";

export const _Prefix = Object.freeze({
  base: GIT_PREFIX,
  blob: _gitAddress(BLOB_TYPE),
  commit: _gitAddress(COMMIT_TYPE),
  submoduleCommit: _gitAddress(SUBMODULE_COMMIT_TYPE),
  tree: _gitAddress(TREE_TYPE),
  treeEntry: _gitAddress(TREE_ENTRY_TYPE),
});

export type BlobAddress = {|
  +type: typeof BLOB_TYPE,
  +hash: Hash,
|};
export type CommitAddress = {|
  +type: typeof COMMIT_TYPE,
  +hash: Hash,
|};
export type SubmoduleCommitAddress = {|
  +type: typeof SUBMODULE_COMMIT_TYPE,
  +submoduleUrl: string,
  +commitHash: Hash,
|};
export type TreeAddress = {|
  +type: typeof TREE_TYPE,
  +hash: Hash,
|};
export type TreeEntryAddress = {|
  +type: typeof TREE_ENTRY_TYPE,
  +treeHash: Hash,
  +name: string,
|};

// A tree entry has contents with one of the following types of
// addresses.
export type TreeEntryContentsAddress =
  | BlobAddress
  | TreeAddress
  | SubmoduleCommitAddress;

export type StructuredAddress =
  | BlobAddress
  | CommitAddress
  | SubmoduleCommitAddress
  | TreeAddress
  | TreeEntryAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = NodeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "BLOB": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: BLOB_TYPE, hash};
    }
    case "COMMIT": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: COMMIT_TYPE, hash};
    }
    case "SUBMODULE_COMMIT": {
      if (rest.length !== 2) throw fail();
      const [submoduleUrl, commitHash] = rest;
      return {type: SUBMODULE_COMMIT_TYPE, submoduleUrl, commitHash};
    }
    case "TREE": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: TREE_TYPE, hash};
    }
    case "TREE_ENTRY": {
      if (rest.length !== 2) throw fail();
      const [treeHash, name] = rest;
      return {type: TREE_ENTRY_TYPE, treeHash, name};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case BLOB_TYPE:
      return NodeAddress.append(_Prefix.blob, x.hash);
    case COMMIT_TYPE:
      return NodeAddress.append(_Prefix.commit, x.hash);
    case SUBMODULE_COMMIT_TYPE:
      return NodeAddress.append(
        _Prefix.submoduleCommit,
        x.submoduleUrl,
        x.commitHash
      );
    case TREE_TYPE:
      return NodeAddress.append(_Prefix.tree, x.hash);
    case TREE_ENTRY_TYPE:
      return NodeAddress.append(_Prefix.treeEntry, x.treeHash, x.name);
    default:
      // eslint-disable-next-line no-unused-expressions
      (x.type: empty);
      throw new Error(`Unexpected type ${x.type}`);
  }
}
