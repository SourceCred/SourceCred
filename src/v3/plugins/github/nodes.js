// @flow

import {NodeAddress, type NodeAddressT} from "../../core/graph";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GITHUB_PREFIX = NodeAddress.fromParts(["sourcecred", "github"]);
function githubAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GITHUB_PREFIX, ...parts);
}

export type RepoAddress = {|
  +type: "REPO",
  +owner: string,
  +name: string,
|};
export type IssueAddress = {|
  +type: "ISSUE",
  +repo: RepoAddress,
  +number: string,
|};
export type PullAddress = {|
  +type: "PULL",
  +repo: RepoAddress,
  +number: string,
|};
export type ReviewAddress = {|
  +type: "REVIEW",
  +pull: PullAddress,
  +fragment: string,
|};
export type CommentAddress = {|
  +type: "COMMENT",
  +parent: IssueAddress | PullAddress | ReviewAddress,
  +fragment: string,
|};
export type UserlikeAddress = {|
  +type: "USERLIKE",
  +login: string,
|};

export type StructuredAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | UserlikeAddress;

// Each of these types has 0 or more "AUTHORS" edges, each of which
// leads to a UserlikeAddress.  Note: It is not true that every
// Authorable has at least one author, as when GitHub accounts are
// deleted, they leave behind posts without authors.
export type AuthorableAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress;

// Each of these types has text content, which means
// it may be the source of a reference to a ReferentAddress.
export type TextContentAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress;

// Each of these types may be referred to by something
// with text content.
export type ReferentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | UserlikeAddress;

// Each of these types is structurally the child of some
// entity with a ParentAddress.
export type ChildAddress =
  | IssueAddress // child of RepoAddress
  | PullAddress // child of RepoAddress
  | ReviewAddress // child of PullAddress
  | CommentAddress; // child of IssueAddress, PullAddress, or ReviewAddress

// Each of these types may have child nodes (see ChildAddress).
export type ParentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GITHUB_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_gh, kind, ...rest] = NodeAddress.toParts(x);
  switch (kind) {
    case "repo": {
      if (rest.length !== 2) {
        throw fail();
      }
      const [owner, name] = rest;
      return {type: "REPO", owner, name};
    }
    case "issue": {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: "REPO", owner, name};
      return {type: "ISSUE", repo, number};
    }
    case "pull": {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: "REPO", owner, name};
      return {type: "PULL", repo, number};
    }
    case "review": {
      if (rest.length !== 4) {
        throw fail();
      }
      const [owner, name, pullNumber, fragment] = rest;
      const repo = {type: "REPO", owner, name};
      const pull = {type: "PULL", repo, number: pullNumber};
      return {type: "REVIEW", pull, fragment};
    }
    case "comment": {
      if (rest.length < 1) {
        throw fail();
      }
      const [subkind, ...subrest] = rest;
      switch (subkind) {
        case "issue": {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, issueNumber, fragment] = subrest;
          const repo = {type: "REPO", owner, name};
          const issue = {type: "ISSUE", repo, number: issueNumber};
          return {type: "COMMENT", parent: issue, fragment};
        }
        case "pull": {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, pullNumber, fragment] = subrest;
          const repo = {type: "REPO", owner, name};
          const pull = {type: "PULL", repo, number: pullNumber};
          return {type: "COMMENT", parent: pull, fragment};
        }
        case "review": {
          if (subrest.length !== 5) {
            throw fail();
          }
          const [owner, name, pullNumber, reviewFragment, fragment] = subrest;
          const repo = {type: "REPO", owner, name};
          const pull = {type: "PULL", repo, number: pullNumber};
          const review = {type: "REVIEW", pull, fragment: reviewFragment};
          return {type: "COMMENT", parent: review, fragment};
        }
        default:
          throw fail();
      }
    }
    case "userlike": {
      if (rest.length !== 1) {
        throw fail();
      }
      const [login] = rest;
      return {type: "USERLIKE", login};
    }
    default:
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case "REPO":
      return githubAddress("repo", x.owner, x.name);
    case "ISSUE":
      return githubAddress("issue", x.repo.owner, x.repo.name, x.number);
    case "PULL":
      return githubAddress("pull", x.repo.owner, x.repo.name, x.number);
    case "REVIEW":
      return githubAddress(
        "review",
        x.pull.repo.owner,
        x.pull.repo.name,
        x.pull.number,
        x.fragment
      );
    case "COMMENT":
      switch (x.parent.type) {
        case "ISSUE":
          return githubAddress(
            "comment",
            "issue",
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.fragment
          );
        case "PULL":
          return githubAddress(
            "comment",
            "pull",
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.fragment
          );
        case "REVIEW":
          return githubAddress(
            "comment",
            "review",
            x.parent.pull.repo.owner,
            x.parent.pull.repo.name,
            x.parent.pull.number,
            x.parent.fragment,
            x.fragment
          );
        default:
          // eslint-disable-next-line no-unused-expressions
          (x.parent.type: empty);
          throw new Error(`Bad comment parent type: ${x.parent.type}`);
      }
    case "USERLIKE":
      return githubAddress("userlike", x.login);
    default:
      // eslint-disable-next-line no-unused-expressions
      (x.type: empty);
      throw new Error(`Unexpected type ${x.type}`);
  }
}
