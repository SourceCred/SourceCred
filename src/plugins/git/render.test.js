// @flow

import {makeRepoId} from "../../core/repoId";
import * as GN from "./nodes";
import {description} from "./render";
import type {Repository} from "./types";

describe("plugins/git/render", () => {
  const exampleHash = "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f";
  const exampleCommit: GN.CommitAddress = Object.freeze({
    type: GN.COMMIT_TYPE,
    hash: exampleHash,
  });
  const exampleRepository: Repository = Object.freeze({
    commits: {
      [exampleHash]: {
        hash: exampleHash,
        shortHash: exampleHash.slice(0, 7),
        summary: "This is an example commit",
        parentHashes: [],
      },
    },
    commitToRepoId: {
      [exampleHash]: {[(makeRepoId("sourcecred", "example-git"): any)]: true},
    },
  });

  it("commit snapshots as expected", () => {
    expect(description(exampleCommit, exampleRepository)).toMatchSnapshot();
  });
  it("logs an error for a commit not in the repository", () => {
    const badCommit = {type: GN.COMMIT_TYPE, hash: "1234"};
    // $ExpectFlowError
    console.error = jest.fn();
    expect(description(badCommit, exampleRepository)).toBe("1234");
    expect(console.error).toHaveBeenCalledWith(
      "Unable to find data for commit 1234"
    );
  });
});
