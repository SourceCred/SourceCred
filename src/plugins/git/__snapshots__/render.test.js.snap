// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/git/render commit snapshots as expected 1`] = `"commit 3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f"`;
